File: go/0142M-环形链表II.go
/*
 * @Author: SukiEva
 * @Date: 2021-10-16 14:43:16
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			for head != fast {
				head = head.Next
				fast = fast.Next
			}
			return head
		}
	}
	return nil
}

File: cpp/0142M-环形链表II.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-13 14:49:56
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *pa = head, *pb = head;
        while (pb != nullptr && pb->next != nullptr) {
            pa = pa->next;
            pb = pb->next->next;
            if (pa == pb) {
                pa = head;
                while (pa != pb) {
                    pa = pa->next;
                    pb = pb->next;
                }
                return pa;
            }
        }
        return nullptr;
    }
};
